[SPARK-10116] [core] XORShiftRandom.hashSeed is random in high bits

squito opened the pull request:
https://issues.apache.org/jira/browse/SPARK-10116

This is really trivial, just happened to notice it -- if `XORShiftRandom.hashSeed` is really supposed to have random bits throughout (as the comment implies), it needs to do something for the conversion to `long`.  

@mengxr @mkolod

squito committed:
XORShiftRandom.hashSeed us random in high bits

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41251 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41251/consoleFull) for   PR 8314 at commit [`148d723`](https://github.com/apache/spark/commit/148d7237b4fc1468111058c163a506e8da5a3fa2).

srowen commented:
LGTM. There are better / faster RNGs in standard libs like commons math. @mengxr is it worth Spark having its own still?

SparkQA commented:
[Test build #41251 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41251/console) for   PR 8314 at commit [`148d723`](https://github.com/apache/spark/commit/148d7237b4fc1468111058c163a506e8da5a3fa2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41251/
Test FAILed.

squito commented:
hmm, some of the errors are just checks against the expected sequence from the rng, which I can update  (though some of these tests probably shouldn't require a "perfect" seed).  But I'm a little perplexed by some of the failures, eg. `Word2Vec`.  I'm not an expert on that part of the code, but unless I've done something wrong here, that really shouldn't be so sensitive to the seed, right?  Though that also seems to have a carefully chosen seed to make things work ...

mengxr commented:
Some unit tests and python doctests do depend on the seed, more or less sensitive. I don't think requiring exact output is that bad because it can at least notify us changes in behavior. In Python, the doctest is used to generate documentation. It is useful to show actual output rather than checking the bounds, e.g., https://github.com/apache/spark/blob/master/python/pyspark/sql/dataframe.py#L459.

There is a trade-off between having meaningful probabilistic bounds vs. keeping unit tests small. For example, in Word2Vec, we can increase the training dataset size to reduce the variance of the model output and hence robust to random seed, but that increases the test time too.

That being said, I can help make those tests less sensitive. Do you mind making JIRAs for each of them?

Regarding @srowen 's question, if adding commons-math3 dependency is not an issue and its RNG performs similarly to the one here. I think we shouldn't maintain our own. However, I'm still a little worried about compatibility issues between commons-math3 releases.

srowen commented:
@mengxr commons math 3.x is already a dependency in core. I don't have benchmarks handy, but my experience with the RNGs is that they're at least "more than fast enough" for any purpose I've had. I don't think the RNGs are changing, and they implement particular RNG processes like Well19937 that should not change over time. The down-side to not using it is simply the higher probability of bugs, like this one, when implementing from scratch.

I think we need to get this fix in in any event. @squito do you need a hand? I think simply loosening the tests is appropriate, even if it means bigger tests.

andrewor14 commented on core/src/test/scala/org/apache/spark/util/random/XORShiftRandomSuite.scala:
```diff
@@ -65,4 +65,19 @@ class XORShiftRandomSuite extends SparkFunSuite with Matchers {
   }
+
+  test ("hashSeed has random bits throughout") {
+    val totalBitCount = (0 until 10).map{seed =>
```
style: `.map { seed =>`

andrewor14 commented on core/src/test/scala/org/apache/spark/util/random/XORShiftRandomSuite.scala:
```diff
@@ -65,4 +65,19 @@ class XORShiftRandomSuite extends SparkFunSuite with Matchers {
     assert(random.nextInt() != 0)
   }
+
+  test ("hashSeed has random bits throughout") {
```
no space `test(`

squito committed:
fix JavaAPISuite

squito committed:
style

squito committed:
increase tolerance (for low count tests)

squito committed:
update magic values

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito commented:
@srowen @mengxr sorry for letting this lay around for so long, somehow I completely forgot to update it.  so I updated the tests with a few magic values, but didn't update the algorithms themselves.  I tried a brute force search over 1000 seeds for `Word2Vec`, but none of them made the test pass.  And it looks like `MultiLayerPerceptronClassifier` just [ignores its seed](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala#L777).

fixing those remaining issues is probably a bit beyond me, but this can't be merged till we get those fixes.  How would you like me to proceed here?

squito commented on core/src/test/scala/org/apache/spark/rdd/PairRDDFunctionsSuite.scala:
```diff
@@ -588,7 +588,7 @@ class PairRDDFunctionsSuite extends SparkFunSuite with SharedSparkContext {
       val stdev = if (withReplacement) math.sqrt(expected) else math.sqrt(expected * p * (1 - p))
       // Very forgiving margin since we're dealing with very small sample sizes most of the time
-      math.abs(actual - expected) <= 6 * stdev
+      math.abs(actual - expected) <= 6 * stdev + 2
```
I thought this was a better way to leave a bigger margin for the low count cases, rather than increasing the multiplier.  As another random point, `4 * stdev + 4` also worked.

srowen replied:
Really, this expression relies upon assuming that the binomial and Poisson distribution are well approximated by a normal distribution. When the expected value is in the 10s or 20s this probably isn't very true. This could be rewritten to properly compute the probability using PoissonDistribution and BinomialDistribution. However I think it would be faster to just make sure that the RDD size is not less than 1000 or so in the tests above. (Also, the parts where it computes the expected count with math.ceil are unnecessary: no reason to require these to be an integer, and they're another source of small errors. Let expected be a Double.)

SparkQA commented:
**[Test build #43988 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43988/consoleFull)** for PR 8314 at commit [`9451cba`](https://github.com/apache/spark/commit/9451cba481c5ce6c76d17ec2d5f34a34e0286737).

SparkQA commented:
**[Test build #43988 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43988/consoleFull)** for PR 8314 at commit [`9451cba`](https://github.com/apache/spark/commit/9451cba481c5ce6c76d17ec2d5f34a34e0286737).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43988/
Test FAILed.

srowen commented on core/src/main/scala/org/apache/spark/util/random/XORShiftRandom.scala:
```diff
@@ -60,9 +60,11 @@ private[spark] class XORShiftRandom(init: Long) extends JavaRandom(init) {
-    MurmurHash3.bytesHash(bytes)
+    val lowBits = MurmurHash3.bytesHash(bytes)
+    val highBits = MurmurHash3.bytesHash(bytes, lowBits)
+    (highBits.toLong << 32) | (lowBits.toLong & 0xFFFFFFFFL)
```
Although normally it might not be 100% valid to make a 64-bit hash out of two 32-bit hashes this way (I'm not clear that reusing the seed doesn't connect the bits in some subtle way), it's certainly a big improvement and probably entirely fine for this purpose. I'd still like to remove XORShiftRandom, but that can be for another day.

SparkQA commented:
**[Test build #1950 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1950/consoleFull)** for PR 8314 at commit [`9451cba`](https://github.com/apache/spark/commit/9451cba481c5ce6c76d17ec2d5f34a34e0286737).

SparkQA commented:
**[Test build #1950 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1950/consoleFull)** for PR 8314 at commit [`9451cba`](https://github.com/apache/spark/commit/9451cba481c5ce6c76d17ec2d5f34a34e0286737).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
@squito I think the test failures are legitimate although they are likely merely due to relying too much on the random sequence before.

squito commented:
Hi @srowen -- I know the test failures are legitimate, but I'm not sure what to do about them.

> I tried a brute force search over 1000 seeds for Word2Vec, but none of them made the test pass. And it looks like MultiLayerPerceptronClassifier just ignores its seed.

I need a bit of guidance on how to fix things -- I dunno if the right solution is looser checks, or just updating the magic values assuming that the current implementation is correct.  (And the right solution might be a bit beyond me abilities at the moment ...)

srowen commented:
For `JavaDataFrameSuite.testSampleBy` I think you can accept any value between 1 and 6 for key 0, and 4 and 9 for key 1. These are not-improbable values given the test -- basically, how many of 33 elements do you choose if choosing with probability 0.1 and 0.2 respectively.

The `Word2Vec` test does look far too tight, I think. The others, I'm not as sure. I think the `StreamingKMeansSuite` just needs more points. Let me see if I can provide a concrete suggestion on these.

squito committed:
update tests

squito commented:
I updated some of the tests -- I think just `Word2Vec` and `MultilayerPerceptron` are remaining.  For `JavaDataFrameSuite.testSampleBy` actually had to widen the range for key 1 up to 11, which I think is the 95%.   `DataFrameSuite` uses a test setup special to sql ... I copped out and just updated the magic values.  And for `StreamingKMeanSuite`, I just needed to update it to choose the centers based on which one was actually closer.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44458 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44458/consoleFull)** for PR 8314 at commit [`134533c`](https://github.com/apache/spark/commit/134533cba3d350c73ba05bb5aaa5681f1580c485).

SparkQA commented:
**[Test build #44458 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44458/consoleFull)** for PR 8314 at commit [`134533c`](https://github.com/apache/spark/commit/134533cba3d350c73ba05bb5aaa5681f1580c485).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44458/
Test FAILed.

srowen commented:
PS it's really on me to get this one working and check up on the tests. I'm back on it today. Thank you for pushing it this far. I want to get this one in.

srowen commented on core/src/test/java/org/apache/spark/JavaAPISuite.java:
```diff
@@ -146,21 +146,29 @@ public void intersection() {
+    long s2 = splits[2].count();
+    Assert.assertTrue(s0 + " not within expected range", s0 > 100 && s0 < 300);
+    Assert.assertTrue(s1 + " not within expected range", s1 > 200 && s0 < 400);
+    Assert.assertTrue(s2 + " not within expected range", s2 > 500 && s2 < 700);
```
Expected values are 200, 300, 500. The ranges are very wide; really I'd do +/- 50. The final range should be a bit bigger; 430-570 is OK.

squito replied:
yes, sorry about that.  I think I was playing around with those ranges with much smaller samples until I decided to just use 1000 elements and didn't think about fixing the checks when they passed.  I'll update.

srowen commented on sql/core/src/test/java/test/org/apache/spark/sql/JavaDataFrameSuite.java:
```diff
@@ -258,7 +258,9 @@ public void testSampleBy() {
+    Assert.assertEquals(0, actual[0].getLong(0));
+    Assert.assertTrue(1 <= actual[0].getLong(1) && actual[0].getLong(1) <= 6);
+    Assert.assertEquals(1, actual[1].getLong(0));
+    Assert.assertTrue(4 <= actual[1].getLong(1) && actual[1].getLong(1) <= 11);
```
Right now the test returns 6 and 11 right?  I'm just estimating good bounds using a binomial cdf, but 0-8 and 2-13 are reasonable ranges for both given the setup (roughly, less than 0.1% chance something outside that range is OK). Does the same need to be applied in `DataFrameStatSuite` below ideally?

squito replied:
makes sense, I'll update those ranges.

Ideally, yes I'd do the same to DataFrameStatSuite -- however the `checkAnswer` function expects an exact match, and adds some extra info about the query plan etc.  I can update that to take a more general comparison function, but seems like overkill?

srowen replied:
If it's more complicated than just checking for the same range you do here, leave it. If it's really the same thing and just a matter of adding a range check, probably worth it.

srowen commented:
@yinxusen I'm trying to help debug the test failure in `Word2VecSuite` added in https://github.com/apache/spark/commit/c9d530e2e5123dbd4fd13fc487c890d6076b24bf#diff-a081b952fe8b6e09a492ef37e157b456 I find that changing the seed at all makes the result quite different and the test fails. However the expected value is computed in a pretty clear way; I'm trying to figure out why a particular seed is required here. Are the word vectors chosen to work for the default seed of 42? 

While looking at it I think we can fix one small thing in `org.apache.spark.mllib.feature.Word2Vec`. The initial random vector (which is what the seed affects) is not quite chosen uniformly:

```
      Array.fill[Float](vocabSize * vectorSize)((initRandom.nextFloat() - 0.5f) / vectorSize)
```
should be more like
```
    Array.fill[Float](vocabSize * vectorSize)((initRandom.nextGaussian().toFloat) / vectorSize)
```

This isn't really the issue but something we could adjust. Of course this also makes the test fail.

If we can't figure this out... I think we could hard-code this test to work with the new seed behavior, on the theory that it's probably a test issue.

srowen commented:
The last failure is in `MultilayerPerceptronClassifierSuite`. The test is backwards in that expected/actual are flipped. It should be

    assert(lrMetrics.confusionMatrix ~== mlpMetrics.confusionMatrix absTol 100)

That is the output ought to look something like

```
[info]   152.0  74.0   121.0  
[info]   65.0   167.0  64.0   
[info]   114.0  76.0   167.0  
```

... which is a little strange since this shows a fairly poor classifier's confusion matrix.

There are 2 seeds in this test and setting them to a range of values succeeds in every case for me. This one might be a matter of picking a different seed? although it is a little funny that in this case the MLP classifier never predicted class 0. But that's a different issue.

I also note that LogisticRegressionSuite uses a regular java.util.Random instead of XORShiftRandom. Might be worth adjusting, unless it causes more failures.

yinxusen commented:
@srowen Yes the word vectors chosen to work for seed 42. But the default value for Word2Vec in ML package has been deleted. So I set the seed as 42 in the test suite.

srowen commented:
@yinxusen OK that explains it. Is there any easy way to work backwards from the output to get the right input vectors? we have a fix here for the seeding process that changes the seed and thus means the test won't pass. Failing this, we can of course hard-code the correct new output for this input, which seems suboptimal.

yinxusen commented:
@srowen Actually the test suite of `Word2Vec` in ML package does not test the real `Word2Vec` ability. IMHO, we can generate the `codes` via the `Word2Vec` in MLlib package directly, other than hard-coding them in the text file. If needed, I can help you changing the test suite.

squito committed:
fix bounds

srowen commented:
@yinxusen it would be great if you could lend a hand and sketch out how this would look in a comment. It would probably save me a bit of digging and trial and error as I haven't actually used the class before. That would be very helpful, and after that, I think we might have an answer to all the test failures that come from changing the seed.

squito committed:
review feedback

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito commented:
I think I've fixed the easy stuff (though I was lazy on `DataFrameStatSuite`, I kept it to just checking the values that happen to come out, rather than using a bound, since the test helpers expect an exact check).   `Word2Vec` and `MultilayerPerceptron` are still to go, but I will leave that up to others.

SparkQA commented:
**[Test build #44938 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44938/consoleFull)** for PR 8314 at commit [`c197f83`](https://github.com/apache/spark/commit/c197f83bdcfddc57f7a6a4d9183c96082cdfa555).

SparkQA commented:
**[Test build #44938 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44938/consoleFull)** for PR 8314 at commit [`c197f83`](https://github.com/apache/spark/commit/c197f83bdcfddc57f7a6a4d9183c96082cdfa555).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44938/
Test FAILed.

squito commented:
@srowen on multilayer perceptron --you see it pass before this change for a range of seeds because the [seed is completely ignored](https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/ann/Layer.scala#L777).

(though maybe you are referring to something else, I'm not entirely sure what you mean by 2 seeds -- I only see one seed in each test ...)

srowen commented:
So for `Word2VecSuite`, a stop-gap to get this passing in order to get this fix in is to change the expected output to match the actual final output (wouldn't mind fixing the nextFloat / nextGaussian thing too) and later sort out how to changes `codes`.

For `MultilayerPerceptronClassiferSuite`, I was referring to the two seeds in the test. One is `11L` to `MultilayerPerceptronClassifier` and one is `42` to `generateMultinomialLogisticInput`. I don't know how much either of them matters. Since the test now fails, something must depend on the new, different seeding; could be the second one? in which case, may just be a case of hunting for a seed that works there, or loosening the test or something.

How do you want to proceed to close this out? hard-coding some new test results seems reasonable in these cases since we're fixing a bug, improving other tests, and not making the others any more brittle.

squito commented:
how about I just hard-code new results for now, and I open two new jiras for fixing those cases properly?

srowen commented:
@squito I tend to agree with that approach on the grounds that the point here is fixing the seeding. Making the tests less hard-coded is a separate issue if it needs to be. Maybe the perceptron test just needs a slightly different seed; maybe @yinxusen has a good answer for updating word2vec, but I don't think we should block this particular fix indefinitely otherwise.

squito committed:
better seed for MultilayerPerceptron; magic values for Word2Vec

squito committed:
Merge branch 'master' into SPARK-10116

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

squito commented:
@srowen good call on the extra seed for multilayer perceptron -- I found one pretty easily for the input data that made the test pass.  Perhaps could still be more general, but at least its no worse than before.  And I created https://issues.apache.org/jira/browse/SPARK-11502 for finding a better solution for Word2Vec.

SparkQA commented:
**[Test build #45025 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45025/consoleFull)** for PR 8314 at commit [`96eb00d`](https://github.com/apache/spark/commit/96eb00dd8f3ea92c5b21a90faba9af4da8671d76).

SparkQA commented:
**[Test build #45025 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45025/consoleFull)** for PR 8314 at commit [`96eb00d`](https://github.com/apache/spark/commit/96eb00dd8f3ea92c5b21a90faba9af4da8671d76).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45025/
Test FAILed.

squito committed:
more magic vals for python tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45046/consoleFull)** for PR 8314 at commit [`e05a035`](https://github.com/apache/spark/commit/e05a035998b72832e6f9c6587b0bf68477f5ff3a).

SparkQA commented:
**[Test build #45046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45046/consoleFull)** for PR 8314 at commit [`e05a035`](https://github.com/apache/spark/commit/e05a035998b72832e6f9c6587b0bf68477f5ff3a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45046/
Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45093 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45093/consoleFull)** for PR 8314 at commit [`e05a035`](https://github.com/apache/spark/commit/e05a035998b72832e6f9c6587b0bf68477f5ff3a).

SparkQA commented:
**[Test build #45093 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45093/consoleFull)** for PR 8314 at commit [`e05a035`](https://github.com/apache/spark/commit/e05a035998b72832e6f9c6587b0bf68477f5ff3a).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45093/
Test FAILed.

squito committed:
more pyspark magic vals

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45101 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45101/consoleFull)** for PR 8314 at commit [`04214d9`](https://github.com/apache/spark/commit/04214d92f25e2bae0bb96cbbdeb086131ebf4faa).

SparkQA commented:
**[Test build #45101 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45101/consoleFull)** for PR 8314 at commit [`04214d9`](https://github.com/apache/spark/commit/04214d92f25e2bae0bb96cbbdeb086131ebf4faa).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45101/
Test FAILed.

srowen commented:
@squito very nice work here. I think the SparkR tests finally need a similar treatment and it's good to go. Thankfully I think the errors are quite plain about what they expect, and they do look like the expected differences you'd see from a seed change.

squito committed:
magic R values

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45120 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45120/consoleFull)** for PR 8314 at commit [`5e35321`](https://github.com/apache/spark/commit/5e35321e7f3856aa60b784609161175a6eebc63e).

SparkQA commented:
**[Test build #45120 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45120/consoleFull)** for PR 8314 at commit [`5e35321`](https://github.com/apache/spark/commit/5e35321e7f3856aa60b784609161175a6eebc63e).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45120/
Test FAILed.

squito committed:
another try at finding the right magic values

squito commented on R/pkg/inst/tests/test_sparkSQL.R:
```diff
@@ -1458,8 +1458,8 @@ test_that("sampleBy() on a DataFrame", {
-  expect_identical(as.list(result[1, ]), list(key = "0", count = 2))
-  expect_identical(as.list(result[2, ]), list(key = "1", count = 10))
+  expect_identical(as.list(result[1, ]), list(key = "0", count = 3))
+  expect_identical(as.list(result[2, ]), list(key = "1", count = 7))
```
this one really confused me ... I thought I could figure out what the right values should be by running the equivalent thing in scala, but I got totally different answers.

```scala
case class Data(val key: String)
val data = (0 to 99).map{x => Data((x % 3).toString)}
val dataDF = sqlContext.createDataFrame(data)
scala> dataDF.stat.sampleBy("key", Map("0" -> 0.1, "1" -> 0.2), 0).groupBy("key").count().show()
+---+-----+
|key|count|
+---+-----+
|  0|    5|
|  1|    5|
+---+-----+
```

but whatever, these vals work :/

mengxr replied:
This is because the default number of partitions is different in Scala/Python/R tests.

squito replied:
ah, great, glad there is a good explanation for this :)

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45143 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45143/consoleFull)** for PR 8314 at commit [`7ce14fa`](https://github.com/apache/spark/commit/7ce14faa75cc42026daeed463f0f4ef56fce2639).

SparkQA commented:
**[Test build #45143 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45143/consoleFull)** for PR 8314 at commit [`7ce14fa`](https://github.com/apache/spark/commit/7ce14faa75cc42026daeed463f0f4ef56fce2639).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45143/
Test PASSed.

srowen commented:
Great let's merge this one to master and 1.6. Not sure about 1.5 given that this is a low-priority fix and it ends up changing test behavior in several places.

avulanov commented:
Could you notify the developer community about such changes in advance and point out to probable problematic places? That would be extremely helpful.

srowen commented:
@avulanov what do you mean in this case? Just that default random behavior may change? That is a given though

avulanov commented:
@srowen Yes, exactly! In my case, the tests based on RNG were failing on amplab's jenkins and at the same time were running correctly in my environment, because my version of Spark was 1 day older than the one on jenkins. Indeed, it worth updating to the latest version all the time :) However, it would be great to have notifications about such changes after which one **must** update.

srowen commented:
Yeah the problem was that the seed processing was actually wrong so had to change. I doubt it will change much. But stochastic behavior isn't part of the APIs so you would want to write tests that don't depend on a particular seed anyway. I dont think this is something special to note.

avulanov commented:
I was thinking about removing the stochastic part from the tests. However, the issue is that I need to test that stochastic initialization of parameters for machine learning does work, i.e. the optimization will converge with such parameters. Could you suggest a better way of doing this as opposed to use the seed?
